frontends/ast/simplify.cc

```diff
@@ -2875,12 +2863,76 @@ bool AstNode::simplify(bool const_fold, bool in_lvalue, int stage, int width_hin
 		{
 			// big case block
 
+			int stride = 1;
+
+			int case_width_hint;
+			bool case_sign_hint;
+			shift_expr->detectSignWidth(case_width_hint, case_sign_hint);
+
+			// Extract (index)*(width) from non_opt_range pattern ((@selfsz@((index)*(width)))+(0)).
+			AstNode *lsb_expr =
+				shift_expr->type == AST_ADD && shift_expr->children[0]->type == AST_SELFSZ &&
+				shift_expr->children[1]->type == AST_CONSTANT && shift_expr->children[1]->integer == 0 ?
+				shift_expr->children[0]->children[0] :
+				shift_expr;
+
+			// Optimization: Extract stride from indexing of two-dimensional packed arrays and
+			// variable slices on the form dst[i*stride +: width] = src.
+			if (lsb_expr->type == AST_MUL &&
+			    (lsb_expr->children[0]->type == AST_CONSTANT ||
+			     lsb_expr->children[1]->type == AST_CONSTANT))
+			{
+				int const_i = lsb_expr->children[1]->type == AST_CONSTANT;
+				stride = (int)lsb_expr->children[const_i]->integer;
+			}
+			else if (member_node)  // Member in packed struct/union
+			{
+				// Clamp chunk to range of member within struct/union.
+				log_assert(!source_offset && !children[0]->id2ast->range_swapped);
+
+				// When the (* nowrshmsk *) attribute is set, a CASE block is generated below
+				// to select the indexed bit slice. When a multirange array is indexed, the
+				// start of each possible slice is separated by the bit stride of the last
+				// index dimension, and we can optimize the CASE block accordingly.
+				// The dimension of the original array expression is saved in the 'integer' field.
+				int dims = children[0]->integer;
+				stride = source_width;
+				for (int dim = 0; dim < dims; dim++) {
+					stride /= get_struct_range_width(member_node, dim);
+				}
+			}
+
+			// For (truncated) i*stride to be within the range of dst, the following must hold:
+			//   i*stride ≡ bitno (mod shift_mod), i.e.
+			//   i*stride = k*shift_mod + bitno
+			//
+			// The Diophantine equation on the form ax + by = c:
+			//   stride*i - shift_mod*k = bitno
+			// has solutions iff c is a multiple of d = gcd(a, b), i.e.
+			//   bitno mod gcd(stride, shift_mod) = 0
+			//
+			// long long is at least 64 bits in C++11
+			int max_bits = case_width_hint;
+			long long max_offset = (1ll << (max_bits - case_sign_hint)) - 1;
+			long long min_offset = case_sign_hint ? -(1ll << (max_bits - 1)) : 0;
+			long long shift_mod = 1ll << (case_width_hint - case_sign_hint);
+			// std::gcd requires C++17
+			// long long bitno_div = std::gcd(stride, shift_mod);
+			long long bitno_div = boost::integer::gcd((long long)stride, shift_mod);
+
 			did_something = true;
 			newNode = new AstNode(AST_CASE, shift_expr);
-			for (int i = 0; i < source_width; i += stride) {
+			for (int i = 1 - result_width; i < source_width; i++) {
+                                // Out of range indexes are handled in genrtlil.cc
 				int start_bit = source_offset + i;
-				int end_bit = std::min(start_bit+result_width,source_width) - 1;
```

Comment:
I think the earlier code here was confused since `start_bit+result_width` had the `source_offset` included but was compared to `source_width` which had not. So this led to incorrect behavior on wires with non-zero `source_offset`, right?

Reply:
> Btw shouldn't we rename all those `source_` variables? It's the target of the assignment, so it's confusing.

I agree! I've been thinking about changing that, however I didn't dare to change the "source_" naming which originates with `source_width`in the shift and mask code.

Should I change all `source_offset` and `source_width` to e.g. `dst_offset` and `dst_width`?

Reply:
Yes, go ahead. I would suggest `wire_offset` and `wire_width`, but really anything will be an improvement it seems.

Reply:
OK, changed to `wire_offset` and `wire_width` now.